whisper.cpp

```diff
@@ -4082,6 +4096,109 @@ static void whisper_sequence_score(
+                    p_swap_vec.emplace_back(i, j);
+                }
+                break;
+            }
+        }
+        if (is_one_copy) {
+            one_copy.insert(i);
+        }
+    }
+
+    // beam-search helpers
+    struct kv_buf {
+        std::vector<uint8_t> k;
+        std::vector<uint8_t> v;
+    };
+
+    static std::vector<kv_buf> kv_bufs(size);
+    static bool is_resized = false;
+
+    if (!is_resized) {
+        for (int i = 0; i < size; i++) {
+            kv_bufs[i].k.resize(ggml_nbytes(src[i].kv_self.k));
+            kv_bufs[i].v.resize(ggml_nbytes(src[i].kv_self.v));
+        }
+        is_resized = true;
+    }
+
+    for (auto & i : two_copy) {
+        // make a copy of KV caches
+        memcpy(kv_bufs[i].k.data(), src[i].kv_self.k->data, kv_bufs[i].k.size());
+        memcpy(kv_bufs[i].v.data(), src[i].kv_self.v->data, kv_bufs[i].v.size());
+    }
+
+    // since two-copy decoder KV caches are protected by kv_bufs, modify them first
+    for (auto & i : two_copy) {
+        // skip the decoder indices that require pointer swapping
+        if (p_swap_set.find(i) != p_swap_set.end()) {continue;}
+
+        if (two_copy.find(view[i]) != two_copy.end()) {
+            // modify KV caches of decoder using data from kv_bufs
+            memcpy(src[i].kv_self.k->data, kv_bufs[view[i]].k.data(), kv_bufs[view[i]].k.size());
+            memcpy(src[i].kv_self.v->data, kv_bufs[view[i]].v.data(), kv_bufs[view[i]].v.size());
+        } else {
+            // modify KV caches of decoder using data from correspond decoder KV caches directly
+            memcpy(src[i].kv_self.k->data, src[view[i]].kv_self.k->data, ggml_nbytes(src[view[i]].kv_self.k));
+            memcpy(src[i].kv_self.v->data, src[view[i]].kv_self.v->data, ggml_nbytes(src[view[i]].kv_self.v));
+        }
+    }
+
+    // then modify one-copy decoder KV caches
+    for (auto & i : one_copy) {
+        // skip the decoder indices that require pointer swapping
+        if (p_swap_set.find(i) != p_swap_set.end()) {continue;}
+
+        if (two_copy.find(view[i]) != two_copy.end()) {
+            // modify KV caches of decoder using data from kv_bufs
+            memcpy(src[i].kv_self.k->data, kv_bufs[view[i]].k.data(), kv_bufs[view[i]].k.size());
+            memcpy(src[i].kv_self.v->data, kv_bufs[view[i]].v.data(), kv_bufs[view[i]].v.size());
+        } else {
+            // modify KV caches of decoder using data from correspond decoder KV caches directly
+            memcpy(src[i].kv_self.k->data, src[view[i]].kv_self.k->data, ggml_nbytes(src[view[i]].kv_self.k));
+            memcpy(src[i].kv_self.v->data, src[view[i]].kv_self.v->data, ggml_nbytes(src[view[i]].kv_self.v));
+        }
+    }
+
+    // swap the pointers
+    for (auto & i : p_swap_vec) {
+        std::swap(src[i.first].kv_self.k->data, src[i.second].kv_self.k->data);
+        std::swap(src[i.first].kv_self.v->data, src[i.second].kv_self.v->data);
+    }
```

Comment:
Clever!

We should swap the entire `kv_self` struct so that things remain consistent:

```c++
std::swap(src[i.first].kv_self, src[i.second].kv_self);
```

Otherwise, we can end up in situation where `decoder[0].kv_self.k->data` points into `decoder[1].kv_self.buf`, which is confusing and error prone.